numba/np/arraymath.py

```diff
@@ -704,75 +704,132 @@ def array_argmin_impl(arry):
+def array_argmax_impl_float(arry):
+    if arry.size == 0:
+        raise ValueError("attempt to get argmax of an empty sequence")
+    for v in arry.flat:
+        max_value = v
+        max_idx = 0
+        break
+    if np.isnan(max_value):
+        return max_idx
+
+    idx = 0
+    for v in arry.flat:
+        if np.isnan(v):
+            return idx
+        if v > max_value:
+            max_value = v
+            max_idx = idx
+        idx += 1
+    return max_idx
 
-    elif isinstance(ty, types.Float):
-        def array_argmax_impl(arry):
-            if arry.size == 0:
-                raise ValueError("attempt to get argmax of an empty sequence")
-            for v in arry.flat:
-                max_value = v
-                max_idx = 0
-                break
-            if np.isnan(max_value):
-                return max_idx
 
-            idx = 0
-            for v in arry.flat:
-                if np.isnan(v):
-                    return idx
-                if v > max_value:
-                    max_value = v
-                    max_idx = idx
-                idx += 1
-            return max_idx
+@register_jitable
+def array_argmax_impl_generic(arry):
+    if arry.size == 0:
+        raise ValueError("attempt to get argmax of an empty sequence")
+    for v in arry.flat:
+        max_value = v
+        max_idx = 0
+        break
+
+    idx = 0
+    for v in arry.flat:
+        if v > max_value:
+            max_value = v
+            max_idx = idx
+        idx += 1
+    return max_idx
+
+
+@overload(np.argmax)
+@overload_method(types.Array, "argmax")
+def array_argmax(arr, axis=None):
+    if isinstance(arr.dtype, (types.NPDatetime, types.NPTimedelta)):
+        flatten_impl = array_argmax_impl_datetime
+    elif isinstance(arr.dtype, types.Float):
+        flatten_impl = array_argmax_impl_float
+    else:
+        flatten_impl = array_argmax_impl_generic
 
+    if is_nonelike(axis):
+        def array_argmax_impl(arr, axis=None):
+            return flatten_impl(arr)
     else:
-        def array_argmax_impl(arry):
-            if arry.size == 0:
-                raise ValueError("attempt to get argmax of an empty sequence")
-            for v in arry.flat:
-                max_value = v
-                max_idx = 0
-                break
+        retty = arr.dtype
 
-            idx = 0
-            for v in arry.flat:
-                if v > max_value:
-                    max_value = v
-                    max_idx = idx
-                idx += 1
-            return max_idx
-    res = context.compile_internal(builder, array_argmax_impl, sig, args)
-    return impl_ret_untracked(context, builder, sig.return_type, res)
+        def array_argmax_impl(arr, axis=None):
+            if arr.ndim > 3:
+                # Transposing currently only works with literals, so can't
+                # calculate new axis order for arbitrary array dimensions.
+                raise ValueError("More than 3 dimensions are not supported.")
```

Comment:
Please could you explain what this restriction is for? This sort of thing seems to work fine:
```python
from numba import njit
import numpy as np

@njit
def foo(a):
    return np.transpose(a, (3, 4, 2, 0, 1))

x = np.arange(np.prod([2, 3, 5, 7, 11])).reshape((11, 7, 5, 3, 2))

got = foo(x)
expected = foo.py_func(x)
np.testing.assert_allclose(got, expected)
```
is the difficulty here perhaps in dealing with tuple manipulation or maybe that the return type depends on some run time value?

Reply:
Seems like there's something going on with method redirection, is that involved here too?

Reply:
Given the following code:

```python
# Make chosen axis the last axis:
            new_axis_order = list(range(arr.ndim))
            for i in range(axis):
                new_axis_order[i] = i
            for i in range(axis, arr.ndim - 1):
                new_axis_order[i] = i + 1
            new_axis_order[arr.ndim - 1] = axis
            transposed_arr = arr.transpose(new_axis_order)
```

I get the following error:

```
Traceback (most recent call last):
  File "/home/itamarst/Devel/numba/numba/core/typeinfer.py", line 154, in propagate
    constraint(typeinfer)
  File "/home/itamarst/Devel/numba/numba/core/typeinfer.py", line 566, in __call__
    self.resolve(typeinfer, typevars, fnty)
  File "/home/itamarst/Devel/numba/numba/core/typeinfer.py", line 586, in resolve
    sig = typeinfer.resolve_call(fnty, pos_args, kw_args)
  File "/home/itamarst/Devel/numba/numba/core/typeinfer.py", line 1538, in resolve_call
    return self.context.resolve_function_type(fnty, pos_args, kw_args)
  File "/home/itamarst/Devel/numba/numba/core/typing/context.py", line 211, in resolve_function_type
    raise last_exception
  File "/home/itamarst/Devel/numba/numba/core/typing/context.py", line 194, in resolve_function_type
    res = self._resolve_user_function_type(func, args, kws)
  File "/home/itamarst/Devel/numba/numba/core/typing/context.py", line 246, in _resolve_user_function_type
    return func.get_call_type(self, args, kws)
  File "/home/itamarst/Devel/numba/numba/core/types/functions.py", line 324, in get_call_type
    failures.raise_error()
  File "/home/itamarst/Devel/numba/numba/core/types/functions.py", line 229, in raise_error
    raise errors.TypingError(self.format())
numba.core.errors.TypingError: No implementation of function Function(<function argmax at 0x7fffe8b84268>) found for signature:
 
 >>> argmax(array(int64, 2d, C), axis=Literal[int](0))
 
There are 2 candidate implementations:
  - Of which 2 did not match due to:
  Overload in function 'array_argmax': File: numba/np/arraymath.py: Line 773.
    With argument(s): '(array(int64, 2d, C), axis=int64)':
   Rejected as the implementation raised a specific error:
     TypingError: Failed in nopython mode pipeline (step: nopython frontend)
   Invalid use of BoundFunction(array.transpose for array(int64, 2d, C)) with parameters (list(int64)<iv=None>)
   
   During: resolving callee type: BoundFunction(array.transpose for array(int64, 2d, C))
   During: typing of call at /home/itamarst/Devel/numba/numba/np/arraymath.py (812)
   
   
   File "numba/np/arraymath.py", line 812:
           def array_argmax_impl(arr, axis=None):
               <source elided>
               new_axis_order[arr.ndim - 1] = axis
               transposed_arr = arr.transpose(new_axis_order)
               ^
   
       Traceback (most recent call last):
         File "/home/itamarst/Devel/numba/numba/core/typeinfer.py", line 154, in propagate
           constraint(typeinfer)
         File "/home/itamarst/Devel/numba/numba/core/typeinfer.py", line 566, in __call__
           self.resolve(typeinfer, typevars, fnty)
         File "/home/itamarst/Devel/numba/numba/core/typeinfer.py", line 615, in resolve
           raise TypingError(msg)
       numba.core.errors.TypingError: Invalid use of BoundFunction(array.transpose for array(int64, 2d, C)) with parameters (list(int64)<iv=None>)
       During: resolving callee type: BoundFunction(array.transpose for array(int64, 2d, C))
       During: typing of call at /home/itamarst/Devel/numba/numba/np/arraymath.py (812)
       The above exception was the direct cause of the following exception:
       Traceback (most recent call last):
         File "/home/itamarst/Devel/numba/numba/core/types/functions.py", line 301, in get_call_type
           sig = temp.apply(nolitargs, nolitkws)
         File "/home/itamarst/Devel/numba/numba/core/typing/templates.py", line 345, in apply
           sig = generic(args, kws)
         File "/home/itamarst/Devel/numba/numba/core/typing/templates.py", line 599, in generic
           disp, new_args = self._get_impl(args, kws)
         File "/home/itamarst/Devel/numba/numba/core/typing/templates.py", line 695, in _get_impl
           impl, args = self._build_impl(cache_key, args, kws)
         File "/home/itamarst/Devel/numba/numba/core/typing/templates.py", line 756, in _build_impl
           disp_type.get_call_type(self.context, args, kws)
         File "/home/itamarst/Devel/numba/numba/core/types/functions.py", line 528, in get_call_type
           self.dispatcher.get_call_template(args, kws)
         File "/home/itamarst/Devel/numba/numba/core/dispatcher.py", line 316, in get_call_template
           self.compile(tuple(args))
         File "/home/itamarst/Devel/numba/numba/core/dispatcher.py", line 911, in compile
           cres = self._compiler.compile(args, return_type)
         File "/home/itamarst/Devel/numba/numba/core/dispatcher.py", line 84, in compile
           raise retval
         File "/home/itamarst/Devel/numba/numba/core/dispatcher.py", line 94, in _compile_cached
           retval = self._compile_core(args, return_type)
         File "/home/itamarst/Devel/numba/numba/core/dispatcher.py", line 112, in _compile_core
           pipeline_class=self.pipeline_class)
         File "/home/itamarst/Devel/numba/numba/core/compiler.py", line 675, in compile_extra
           return pipeline.compile_extra(func)
         File "/home/itamarst/Devel/numba/numba/core/compiler.py", line 419, in compile_extra
           return self._compile_bytecode()
         File "/home/itamarst/Devel/numba/numba/core/compiler.py", line 483, in _compile_bytecode
           return self._compile_core()
         File "/home/itamarst/Devel/numba/numba/core/compiler.py", line 462, in _compile_core
           raise e
         File "/home/itamarst/Devel/numba/numba/core/compiler.py", line 453, in _compile_core
           pm.run(self.state)
         File "/home/itamarst/Devel/numba/numba/core/compiler_machinery.py", line 339, in run
           raise patched_exception
         File "/home/itamarst/Devel/numba/numba/core/compiler_machinery.py", line 330, in run
           self._runPass(idx, pass_inst, state)
         File "/home/itamarst/Devel/numba/numba/core/compiler_lock.py", line 35, in _acquire_compile_lock
           return func(*args, **kwargs)
         File "/home/itamarst/Devel/numba/numba/core/compiler_machinery.py", line 289, in _runPass
           mutated |= check(pss.run_pass, internal_state)
         File "/home/itamarst/Devel/numba/numba/core/compiler_machinery.py", line 262, in check
           mangled = func(compiler_state)
         File "/home/itamarst/Devel/numba/numba/core/typed_passes.py", line 110, in run_pass
           raise_errors=self._raise_errors)
         File "/home/itamarst/Devel/numba/numba/core/typed_passes.py", line 82, in type_inference_stage
           errs = infer.propagate(raise_errors=raise_errors)
         File "/home/itamarst/Devel/numba/numba/core/typeinfer.py", line 1071, in propagate
           raise errors[0]
       numba.core.errors.TypingError: Failed in nopython mode pipeline (step: nopython frontend)
       Invalid use of BoundFunction(array.transpose for array(int64, 2d, C)) with parameters (list(int64)<iv=None>)
       During: resolving callee type: BoundFunction(array.transpose for array(int64, 2d, C))
       During: typing of call at /home/itamarst/Devel/numba/numba/np/arraymath.py (812)
       File "numba/np/arraymath.py", line 812:
               def array_argmax_impl(arr, axis=None):
                   <source elided>
                   new_axis_order[arr.ndim - 1] = axis
                   transposed_arr = arr.transpose(new_axis_order)
                   ^
  raised from /home/itamarst/Devel/numba/numba/core/typeinfer.py:1071

During: resolving callee type: Function(<function argmax at 0x7fffe8b84268>)
During: typing of call at /home/itamarst/Devel/numba/numba/tests/test_array_reductions.py (955)


The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/itamarst/Devel/numba/numba/tests/test_array_reductions.py", line 962, in test_argmax_axis_1d_2d_3d
    self.assertPreciseEqual(cfunc.py_func(arr), cfunc(arr))
  File "/home/itamarst/Devel/numba/numba/core/dispatcher.py", line 421, in _compile_for_args
    error_rewrite(e, 'typing')
  File "/home/itamarst/Devel/numba/numba/core/dispatcher.py", line 360, in error_rewrite
    raise e
  File "/home/itamarst/Devel/numba/numba/core/dispatcher.py", line 373, in _compile_for_args
    return_val = self.compile(tuple(argtypes))
  File "/home/itamarst/Devel/numba/numba/core/dispatcher.py", line 911, in compile
    cres = self._compiler.compile(args, return_type)
  File "/home/itamarst/Devel/numba/numba/core/dispatcher.py", line 84, in compile
    raise retval
  File "/home/itamarst/Devel/numba/numba/core/dispatcher.py", line 94, in _compile_cached
    retval = self._compile_core(args, return_type)
  File "/home/itamarst/Devel/numba/numba/core/dispatcher.py", line 112, in _compile_core
    pipeline_class=self.pipeline_class)
  File "/home/itamarst/Devel/numba/numba/core/compiler.py", line 675, in compile_extra
    return pipeline.compile_extra(func)
  File "/home/itamarst/Devel/numba/numba/core/compiler.py", line 419, in compile_extra
    return self._compile_bytecode()
  File "/home/itamarst/Devel/numba/numba/core/compiler.py", line 483, in _compile_bytecode
    return self._compile_core()
  File "/home/itamarst/Devel/numba/numba/core/compiler.py", line 462, in _compile_core
    raise e
  File "/home/itamarst/Devel/numba/numba/core/compiler.py", line 453, in _compile_core
    pm.run(self.state)
  File "/home/itamarst/Devel/numba/numba/core/compiler_machinery.py", line 339, in run
    raise patched_exception
  File "/home/itamarst/Devel/numba/numba/core/compiler_machinery.py", line 330, in run
    self._runPass(idx, pass_inst, state)
  File "/home/itamarst/Devel/numba/numba/core/compiler_lock.py", line 35, in _acquire_compile_lock
    return func(*args, **kwargs)
  File "/home/itamarst/Devel/numba/numba/core/compiler_machinery.py", line 289, in _runPass
    mutated |= check(pss.run_pass, internal_state)
  File "/home/itamarst/Devel/numba/numba/core/compiler_machinery.py", line 262, in check
    mangled = func(compiler_state)
  File "/home/itamarst/Devel/numba/numba/core/typed_passes.py", line 110, in run_pass
    raise_errors=self._raise_errors)
  File "/home/itamarst/Devel/numba/numba/core/typed_passes.py", line 82, in type_inference_stage
    errs = infer.propagate(raise_errors=raise_errors)
  File "/home/itamarst/Devel/numba/numba/core/typeinfer.py", line 1071, in propagate
    raise errors[0]
numba.core.errors.TypingError: Failed in nopython mode pipeline (step: nopython frontend)
No implementation of function Function(<function argmax at 0x7fffe8b84268>) found for signature:
 
 >>> argmax(array(int64, 2d, C), axis=Literal[int](0))
 
There are 2 candidate implementations:
  - Of which 2 did not match due to:
  Overload in function 'array_argmax': File: numba/np/arraymath.py: Line 773.
    With argument(s): '(array(int64, 2d, C), axis=int64)':
   Rejected as the implementation raised a specific error:
     TypingError: Failed in nopython mode pipeline (step: nopython frontend)
   Invalid use of BoundFunction(array.transpose for array(int64, 2d, C)) with parameters (list(int64)<iv=None>)
   
   During: resolving callee type: BoundFunction(array.transpose for array(int64, 2d, C))
   During: typing of call at /home/itamarst/Devel/numba/numba/np/arraymath.py (812)
   
   
   File "numba/np/arraymath.py", line 812:
           def array_argmax_impl(arr, axis=None):
               <source elided>
               new_axis_order[arr.ndim - 1] = axis
               transposed_arr = arr.transpose(new_axis_order)
               ^
   
       Traceback (most recent call last):
         File "/home/itamarst/Devel/numba/numba/core/typeinfer.py", line 154, in propagate
           constraint(typeinfer)
         File "/home/itamarst/Devel/numba/numba/core/typeinfer.py", line 566, in __call__
           self.resolve(typeinfer, typevars, fnty)
         File "/home/itamarst/Devel/numba/numba/core/typeinfer.py", line 615, in resolve
           raise TypingError(msg)
       numba.core.errors.TypingError: Invalid use of BoundFunction(array.transpose for array(int64, 2d, C)) with parameters (list(int64)<iv=None>)
       During: resolving callee type: BoundFunction(array.transpose for array(int64, 2d, C))
       During: typing of call at /home/itamarst/Devel/numba/numba/np/arraymath.py (812)
       The above exception was the direct cause of the following exception:
       Traceback (most recent call last):
         File "/home/itamarst/Devel/numba/numba/core/types/functions.py", line 301, in get_call_type
           sig = temp.apply(nolitargs, nolitkws)
         File "/home/itamarst/Devel/numba/numba/core/typing/templates.py", line 345, in apply
           sig = generic(args, kws)
         File "/home/itamarst/Devel/numba/numba/core/typing/templates.py", line 599, in generic
           disp, new_args = self._get_impl(args, kws)
         File "/home/itamarst/Devel/numba/numba/core/typing/templates.py", line 695, in _get_impl
           impl, args = self._build_impl(cache_key, args, kws)
         File "/home/itamarst/Devel/numba/numba/core/typing/templates.py", line 756, in _build_impl
           disp_type.get_call_type(self.context, args, kws)
         File "/home/itamarst/Devel/numba/numba/core/types/functions.py", line 528, in get_call_type
           self.dispatcher.get_call_template(args, kws)
         File "/home/itamarst/Devel/numba/numba/core/dispatcher.py", line 316, in get_call_template
           self.compile(tuple(args))
         File "/home/itamarst/Devel/numba/numba/core/dispatcher.py", line 911, in compile
           cres = self._compiler.compile(args, return_type)
         File "/home/itamarst/Devel/numba/numba/core/dispatcher.py", line 84, in compile
           raise retval
         File "/home/itamarst/Devel/numba/numba/core/dispatcher.py", line 94, in _compile_cached
           retval = self._compile_core(args, return_type)
         File "/home/itamarst/Devel/numba/numba/core/dispatcher.py", line 112, in _compile_core
           pipeline_class=self.pipeline_class)
         File "/home/itamarst/Devel/numba/numba/core/compiler.py", line 675, in compile_extra
           return pipeline.compile_extra(func)
         File "/home/itamarst/Devel/numba/numba/core/compiler.py", line 419, in compile_extra
           return self._compile_bytecode()
         File "/home/itamarst/Devel/numba/numba/core/compiler.py", line 483, in _compile_bytecode
           return self._compile_core()
         File "/home/itamarst/Devel/numba/numba/core/compiler.py", line 462, in _compile_core
           raise e
         File "/home/itamarst/Devel/numba/numba/core/compiler.py", line 453, in _compile_core
           pm.run(self.state)
         File "/home/itamarst/Devel/numba/numba/core/compiler_machinery.py", line 339, in run
           raise patched_exception
         File "/home/itamarst/Devel/numba/numba/core/compiler_machinery.py", line 330, in run
           self._runPass(idx, pass_inst, state)
         File "/home/itamarst/Devel/numba/numba/core/compiler_lock.py", line 35, in _acquire_compile_lock
           return func(*args, **kwargs)
         File "/home/itamarst/Devel/numba/numba/core/compiler_machinery.py", line 289, in _runPass
           mutated |= check(pss.run_pass, internal_state)
         File "/home/itamarst/Devel/numba/numba/core/compiler_machinery.py", line 262, in check
           mangled = func(compiler_state)
         File "/home/itamarst/Devel/numba/numba/core/typed_passes.py", line 110, in run_pass
           raise_errors=self._raise_errors)
         File "/home/itamarst/Devel/numba/numba/core/typed_passes.py", line 82, in type_inference_stage
           errs = infer.propagate(raise_errors=raise_errors)
         File "/home/itamarst/Devel/numba/numba/core/typeinfer.py", line 1071, in propagate
           raise errors[0]
       numba.core.errors.TypingError: Failed in nopython mode pipeline (step: nopython frontend)
       Invalid use of BoundFunction(array.transpose for array(int64, 2d, C)) with parameters (list(int64)<iv=None>)
       During: resolving callee type: BoundFunction(array.transpose for array(int64, 2d, C))
       During: typing of call at /home/itamarst/Devel/numba/numba/np/arraymath.py (812)
       File "numba/np/arraymath.py", line 812:
               def array_argmax_impl(arr, axis=None):
                   <source elided>
                   new_axis_order[arr.ndim - 1] = axis
                   transposed_arr = arr.transpose(new_axis_order)
                   ^
  raised from /home/itamarst/Devel/numba/numba/core/typeinfer.py:1071

During: resolving callee type: Function(<function argmax at 0x7fffe8b84268>)
During: typing of call at /home/itamarst/Devel/numba/numba/tests/test_array_reductions.py (955)


File "numba/tests/test_array_reductions.py", line 955:
    def test_argmax_axis_1d_2d_3d(self):
        <source elided>
            py_functions = [
                lambda a, _axis=axis: np.argmax(a, axis=_axis)
                ^
```

Reply:
I am not sure how to fix the above.

Reply:
Here is an isolated reproducer of the problem of trying to create the new axis order programmatically, instead of with hardcoding: https://gist.github.com/itamarst/ebdef3a06a97d409a91a55c818196447

Reply:
Thanks for creating a reproducer. I think the underlying issue here is that `array.transpose` expects a tuple and not a list. I imagine that this is in part to help static analysis, tuples are immutable so it's safe for their type to carry their size. This means that a transpose can easily be verified as potentially valid if the tuple size is the same as the array dimensionality. In the case of providing a list, I think this is technically possible but there would need to be a run time check to ensure that the list is the right size.

Given the actual implementation of `argmax` presented herein, is `new_axis_order` a predictable mutation of `range(arr.ndim)` (I'm fairly sure it is)? In which case I think there could be a "trick" to support higher dimensionality.

Reply:
The logic is "take tuple(range(arr.ndim)) and move the given axis so it's last". So e.g. for ndim 3, axis 1:  `(0, 1, 2) -> (0, 2, 1)`. So it's predictable but it depends on the axis input.

Reply:
I think this may permit the current restriction to be removed: https://github.com/stuartarchibald/numba/commit/c8584986b2fc059450b816614438f6d78b64acf2